web/text_layer_builder.js

```diff
@@ -30,378 +30,376 @@ var EXPAND_DIVS_TIMEOUT = 300; // ms
-        offset: undefined,
+      let match = {
+        begin: {
+          divIdx: i,
+          offset: matchIdx - iIndex,
+        },
       };
 
-      function beginText(begin, className) {
-        var divIdx = begin.divIdx;
-        textDivs[divIdx].textContent = '';
-        appendTextToDiv(divIdx, 0, begin.offset, className);
+      // Calculate the end position.
+      if (matchesLength) { // Multiterm search.
+        matchIdx += matchesLength[m];
+      } else { // Phrase search.
+        matchIdx += queryLen;
       }
 
-      function appendTextToDiv(divIdx, fromOffset, toOffset, className) {
-        var div = textDivs[divIdx];
-        var content =
-          textContentItemsStr[divIdx].substring(fromOffset, toOffset);
-        var node = document.createTextNode(content);
-        if (className) {
-          var span = document.createElement('span');
-          span.className = className;
-          span.appendChild(node);
-          div.appendChild(span);
-          return;
-        }
-        div.appendChild(node);
+      // Somewhat the same array as above, but use > instead of >= to get
+      // the end position right.
+      while (i !== end && matchIdx >
+             (iIndex + textContentItemsStr[i].length)) {
+        iIndex += textContentItemsStr[i].length;
+        i++;
       }
 
-      var i0 = selectedMatchIdx, i1 = i0 + 1;
-      if (highlightAll) {
-        i0 = 0;
-        i1 = matches.length;
-      } else if (!isSelectedPage) {
-        // Not highlighting all and this isn't the selected page, so do nothing.
-        return;
-      }
+      match.end = {
+        divIdx: i,
+        offset: matchIdx - iIndex,
+      };
+      ret.push(match);
+    }
 
-      for (var i = i0; i < i1; i++) {
-        var match = matches[i];
-        var begin = match.begin;
-        var end = match.end;
-        var isSelected = (isSelectedPage && i === selectedMatchIdx);
-        var highlightSuffix = (isSelected ? ' selected' : '');
+    return ret;
+  }
 
-        if (this.findController) {
-          this.findController.updateMatchPosition(pageIdx, i, textDivs,
-                                                  begin.divIdx);
-        }
+  renderMatches(matches) {
+    // Early exit if there is nothing to render.
+    if (matches.length === 0) {
+      return;
+    }
+
+    let textContentItemsStr = this.textContentItemsStr;
+    let textDivs = this.textDivs;
+    let prevEnd = null;
+    let pageIdx = this.pageIdx;
+    let isSelectedPage = (this.findController === null ?
+      false : (pageIdx === this.findController.selected.pageIdx));
+    let selectedMatchIdx = (this.findController === null ?
+                            -1 : this.findController.selected.matchIdx);
+    let highlightAll = (this.findController === null ?
+                        false : this.findController.state.highlightAll);
+    let infinity = {
+      divIdx: -1,
+      offset: undefined,
+    };
+
+    let beginText = function(begin, className) {
```

Comment:
I decided to revert to the original function declarations to reduce the size of the change and because, as you mentioned, the function expression don't add much here.